### PR TITLE
MM-51085: add source and staging models for commonly used Salesforce data

### PR DESCRIPTION
**Pull request by ifoukarakis**
#### Summary

- [x] Add salesforce sources.
- [x] Add staging models for most commonly used tables.
- [x] Add model definitions and basic tests. Includes basic documentation and primary key tests as PR is already pretty big already. Entity documentation based (mostly) on [Salesforce glossary](https://help.salesforce.com/s/articleView?id=sf.glossary.htm&type=5).

Code is (mostly) auto-generated by [codegen](https://github.com/dbt-labs/dbt-codegen). The following changes were done manually:
- Column renaming
- Grouping columns (i.e. foreign keys, contact information, company context, pipeline status, metadata etc).
- Removing stitch related columns
- Docs & tests

I'd also love some feedback on whether we should keep deleted objects or not. Existing models (in the old project) do ignore them early on the process. 

Follow up PRs can improve the code by:
- More tests, especially referential integrity tests.
- Adding more documentation.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-51085

### CODE REVIEW

File: transform/mattermost-analytics/models/staging/salesforce/_salesforce__models.yml [hunk 1]
```diff
@@ -0,0 +1,1661 @@
...
+
+      - name: total_amount_total_price__c
+        description: ""
+
+      - name: total_price_with_annualized_expansion__c
+        description: ""
+
+  - name: stg_salesforce__user
+    description: | 
+      A user is anyone who logs in to Salesforce.
```
Comment (catalintomai) [inline transform/mattermost-analytics/models/staging/salesforce/_salesforce__models.yml:1577]: is this really a valid description? Not sure why we'd need to handle "users that log into SFDC" ??

Comment (ifoukarakis) [reply transform/mattermost-analytics/models/staging/salesforce/_salesforce__models.yml:1577]: > is this really a valid description? Not sure why we'd need to handle "users that log into SFDC" ??

It actually is. Perhaps we can change the wording. Here's a different one: https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_user.htm

This model is used in the old project as part of ARR models.

Comment (catalintomai) [reply transform/mattermost-analytics/models/staging/salesforce/_salesforce__models.yml:1577]: Yes, I'd think the second one is more appropriate? I would not equate "Represents a user in your organization" with "A user is anyone who logs in to Salesforce." - I assume there are plenty of MM users that never logged-in into SFDC.

Comment (ifoukarakis) [reply transform/mattermost-analytics/models/staging/salesforce/_salesforce__models.yml]: Done!

File: transform/mattermost-analytics/models/staging/salesforce/stg_salesforce__campaign.sql [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+renamed as (
+
+    select
+        id as campaign_id,
+        actualcost as actual_cost,
+        amountallopportunities as amount_all_opportunities,
+        amountwonopportunities as amount_won_opportunities,
+        budgetedcost as budgeted_cost,
+        createdbyid as created_by_id,
+        createddate as created_date,
```
[the remark this answers is not in the document]
Comment (ifoukarakis) [reply transform/mattermost-analytics/models/staging/salesforce/stg_salesforce__campaign.sql]: > Lmk if you need/want my help - we can split the tables and scrub them one by one. Unless you have a query that does the checking/identification = )

Done!

Wherever the column name ended with `date`, it was replaced with `_at`. For example `createdate` was converted to `create_at`. Table definitions from snowflake were used to identify such columns.

Columns were also grouped together by context (i.e. primary key, followed by foreign keys, then contact info, company info etc). This should make related columns appear together when running `select *`, but also make it easier understand relationships between models/tables.

Note that this wasn't done for any column with `__c` postfix, as they seem to be "pass-through" columns. If needed, we can do a second pass on this.

File: transform/mattermost-analytics/models/staging/salesforce/stg_salesforce__opportunity.sql [hunk 3]
```diff
@@ -9,29 +9,39 @@ renamed as (
...
+        -- Foreign keys
         accountid as account_id,
-        amount,
         campaignid as campaign_id,
-        closedate as close_date,
         contractid as contract_id,
         createdbyid as created_by_id,
-        createddate as created_date,
+        lastmodifiedbyid as last_modified_by_id,
+        ownerid as onwer_id,
```
Comment (catalintomai) [inline transform/mattermost-analytics/models/staging/salesforce/stg_salesforce__opportunity.sql:19]: onwer->owner

File: transform/mattermost-analytics/models/staging/salesforce/stg_salesforce__contact.sql [hunk 4]
```diff
@@ -9,21 +9,27 @@ renamed as (
...
         id as contact_id,
+
+        -- Foreign keys
         accountid as account_id,
         createdbyid as created_by_id,
-        createddate as created_date,
-        donotcall as do_not_call,
-        email,
+        lastmodifiedbyid as last_modified_by_id,
+        ownerid as onwer_id,
```
Comment (catalintomai) [inline transform/mattermost-analytics/models/staging/salesforce/stg_salesforce__contact.sql:17]: onwer->owner

**Review by catalintomai — APPROVED** [review]

**Pull request merged**